### ISSUE TITLE
App's cache size can get unreasonably high when uploads fail

### ISSUE THREAD

**Issue by jklausa** [wordpress-mobile/WordPress-iOS#12306]
I noticed the other day that the WPiOS on my phone was taking an unreasonable amount of space on device:

![Image from iOS](https://user-images.githubusercontent.com/1594081/62984083-11ee8f80-be32-11e9-8a1b-593b384efd0b.png)

despite the app itself reporting it takes up almost no storage: 
![Image from iOS (1)](https://user-images.githubusercontent.com/1594081/62984105-2763b980-be32-11e9-933f-0ffb5017f041.png)

I decided to investigate and found some truly... peculiar things.

It turned out that almost all of the data was being taken up by an Alamofire cache:

```lldb
(lldb) po fileManager.allocatedSizeOf(directoryURL:temporary.appendingPathComponent("org.alamofire.manager").appendingPathComponent("multipart.form.data"))
14275268608
```
(14275268608 bytes is ~13 gigs)

I then wanted to see the breakdown of what _files_ take up so much space:

```
(lldb) po fileManager.contentsOfDirectory(at: temporary.appendingPathComponent("org.alamofire.manager").appendingPathComponent("multipart.form.data"), includingPropertiesForKeys: nil, options: []).map { return ($0.lastPathComponent, $0.fileSize) }
▿ 46 elements
  ▿ 0 : 2 elements
    - .0 : "423CC996-2EB2-4DDE-AFA4-0921AE6879EF"
    ▿ .1 : Optional<Int64>
      - some : 322242705
  ▿ 1 : 2 elements
    - .0 : "303E785A-1129-430E-818C-1F9957068D2E"
    ▿ .1 : Optional<Int64>
      - some : 258097297
  ▿ 2 : 2 elements
    - .0 : "11D32147-0461-499F-B057-D470FDCDAAA0"
    ▿ .1 : Optional<Int64>
      - some : 335202705
  ▿ 3 : 2 elements
    - .0 : "3F3CC2F0-F6A6-4B9E-8A0D-7349B2F121C1"
    ▿ .1 : Optional<Int64>
      - some : 365084873
  ▿ 4 : 2 elements
    - .0 : "E831F246-8D9A-4DD0-99F0-5287B1B12B61"
    ▿ .1 : Optional<Int64>
      - some : 147545903
  ▿ 5 : 2 elements
    - .0 : "ABFD9537-F0A9-479F-9A5A-466691EA7601"
    ▿ .1 : Optional<Int64>
      - some : 365084873
  ▿ 6 : 2 elements
    - .0 : "8D883614-B4FB-4925-8115-17E9F3B99B49"
    ▿ .1 : Optional<Int64>
      - some : 365084873
  ▿ 7 : 2 elements
    - .0 : "6E7FCB96-316A-41AE-91A5-D0ED0B9D0713"
    ▿ .1 : Optional<Int64>
      - some : 365084873
  ▿ 8 : 2 elements
    - .0 : "5B6FDB66-9F75-4949-A359-DAD2F69587D1"
    ▿ .1 : Optional<Int64>
      - some : 147545903
  ▿ 9 : 2 elements
    - .0 : "0F46659E-0560-4444-8314-2A91D6E24ADE"
    ▿ .1 : Optional<Int64>
      - some : 377276707
  ▿ 10 : 2 elements
    - .0 : "FAB77FCB-B133-4669-B173-2FC757319F07"
    ▿ .1 : Optional<Int64>
      - some : 147545903
  ▿ 11 : 2 elements
    - .0 : "EF33CD55-E521-46EF-B818-F22BC4BCD2BA"
    ▿ .1 : Optional<Int64>
      - some : 335202705
  ▿ 12 : 2 elements
    - .0 : "0CFEBA1B-F3DA-409B-8929-7A09B87D9D4B"
    ▿ .1 : Optional<Int64>
      - some : 147545903
  ▿ 13 : 2 elements
    - .0 : "97994051-571E-4D82-BF41-3B2F9AD85E58"
    ▿ .1 : Optional<Int64>
      - some : 377276707
  ▿ 14 : 2 elements
    - .0 : "57FBDAB4-D3BE-4439-9960-C24B40E4D389"
    ▿ .1 : Optional<Int64>
      - some : 351698519
  ▿ 15 : 2 elements
    - .0 : "D9567EF0-C799-44BB-B3DC-1885D193D21E"
    ▿ .1 : Optional<Int64>
      - some : 377276707
  ▿ 16 : 2 elements
    - .0 : "2EE412FE-07E8-49BB-8B83-EB6555AD494E"
    ▿ .1 : Optional<Int64>
      - some : 247087249
  ▿ 17 : 2 elements
    - .0 : "A8EDCE71-A524-4690-8BF8-AB86F5673CB2"
    ▿ .1 : Optional<Int64>
      - some : 351698519
  ▿ 18 : 2 elements
    - .0 : "DB44DBAE-C1FE-441A-827E-3D1F78AA7FB7"
    ▿ .1 : Optional<Int64>
      - some : 147545903
  ▿ 19 : 2 elements
    - .0 : "64316C16-113F-4E0E-BBE2-CF4C5B8D2BC8"
    ▿ .1 : Optional<Int64>
      - some : 258097297
  ▿ 20 : 2 elements
    - .0 : "F755B1DE-4404-4F33-8381-BB926D73B9F3"
    ▿ .1 : Optional<Int64>
      - some : 351698519
  ▿ 21 : 2 elements
    - .0 : "D01502A7-9A54-4129-B952-1A5BEF397DAE"
    ▿ .1 : Optional<Int64>
      - some : 365084873
  ▿ 22 : 2 elements
    - .0 : "3AD34C25-99ED-4E62-A232-8D350445CBE4"
    ▿ .1 : Optional<Int64>
      - some : 298578065
  ▿ 23 : 2 elements
    - .0 : "1D423734-FC3C-440B-A7EC-9681F2E76CC9"
    ▿ .1 : Optional<Int64>
      - some : 377276707
  ▿ 24 : 2 elements
    - .0 : "46C47C3B-E470-469A-A192-4CCF2A97F656"
    ▿ .1 : Optional<Int64>
      - some : 365084873
  ▿ 25 : 2 elements
    - .0 : "85A3ED0D-297A-40BE-8EBC-D3ACAEB5D3C0"
    ▿ .1 : Optional<Int64>
      - some : 350959761
  ▿ 26 : 2 elements
    - .0 : "4DCD6122-775B-4DF5-9ED1-BD0D33FFB4C5"
    ▿ .1 : Optional<Int64>
      - some : 334284945
  ▿ 27 : 2 elements
    - .0 : "E424578B-9404-4EFD-A96E-5612F34D08E0"
    ▿ .1 : Optional<Int64>
      - some : 147545903
  ▿ 28 : 2 elements
    - .0 : "75A84716-BC4D-46E0-8CAB-A593A4D9EF74"
    ▿ .1 : Optional<Int64>
      - some : 377276707
  ▿ 29 : 2 elements
    - .0 : "5A686F8C-B80D-4FBE-88A4-89EBE6916D41"
    ▿ .1 : Optional<Int64>
      - some : 147545903
  ▿ 30 : 2 elements
    - .0 : "5719A34E-CCA3-40F8-8144-3A170C5764CA"
    ▿ .1 : Optional<Int64>
      - some : 335202705
  ▿ 31 : 2 elements
    - .0 : "94C43243-3202-44C2-B840-154A611071E3"
    ▿ .1 : Optional<Int64>
      - some : 335202705
  ▿ 32 : 2 elements
    - .0 : "D4640ED0-5DAD-4B59-BF19-CBF67BCB18B8"
    ▿ .1 : Optional<Int64>
      - some : 377276707
  ▿ 33 : 2 elements
    - .0 : "D66AE8B8-B167-4861-9102-0CE145352E59"
    ▿ .1 : Optional<Int64>
      - some : 365084873
  ▿ 34 : 2 elements
    - .0 : "74523F0F-A0F6-4D52-8EFA-4C416E16931F"
    ▿ .1 : Optional<Int64>
      - some : 351698519
  ▿ 35 : 2 elements
    - .0 : "4453E05D-AD8B-4107-B34D-8A720E15E652"
    ▿ .1 : Optional<Int64>
      - some : 298578065
  ▿ 36 : 2 elements
    - .0 : "6D9FE416-D81F-4A86-BEBD-4377B5B20287"
    ▿ .1 : Optional<Int64>
      - some : 302018705
  ▿ 37 : 2 elements
    - .0 : "CB740D39-BD3E-4D63-8B4B-6540DC4DB0AE"
    ▿ .1 : Optional<Int64>
      - some : 365084873
  ▿ 38 : 2 elements
    - .0 : "050AE78D-DEDD-481D-B7F8-147F7FA7206D"
    ▿ .1 : Optional<Int64>
      - some : 333897873
  ▿ 39 : 2 elements
    - .0 : "0EFFA3EB-AE97-495D-8E07-C480DC838DCC"
    ▿ .1 : Optional<Int64>
      - some : 351698519
  ▿ 40 : 2 elements
    - .0 : "877BD6FA-0828-452E-811D-986ACDD4593E"
    ▿ .1 : Optional<Int64>
      - some : 247038097
  ▿ 41 : 2 elements
    - .0 : "1BF0693A-9C8D-410E-BA9D-1D44068B9008"
    ▿ .1 : Optional<Int64>
      - some : 334289041
  ▿ 42 : 2 elements
    - .0 : "FFFCA1F0-7AE2-427E-BF30-C22FC5FBB023"
    ▿ .1 : Optional<Int64>
      - some : 335202705
  ▿ 43 : 2 elements
    - .0 : "4ECE2E40-AAA0-4CFC-BB84-DA368E92F738"
    ▿ .1 : Optional<Int64>
      - some : 335202705
  ▿ 44 : 2 elements
    - .0 : "A853E283-3157-4C5B-A6F7-3CDED167AF82"
    ▿ .1 : Optional<Int64>
      - some : 351698519
  ▿ 45 : 2 elements
    - .0 : "55278161-055B-4203-8C55-94E21EDEA8F2"
    ▿ .1 : Optional<Int64>
      - some : 304215185
```

The suspicious part here is that some of the filesizes here are repeating — I _think_ this might be because I was trying out to reproduce https://github.com/wordpress-mobile/WordPress-iOS/issues/12282, and trying to upload some of the videos off of my phone?

Perhaps then, when the upload failed for whatever reason, Alamofire didn't clean up those files properly? We should investigate under what conditions that might happen and see if we can prevent/detect those.

**Comment by jklausa** [comment]
Actually, one more suggestion maybe — maybe we should include the size of `tmp` directory in the cache size displayed in the Settings and/or we should also clear it when the user taps on "Clear Device Media Cache"?

@frosty what do you think?

**Comment by frosty** [comment]
Good catch! I just did a test here and can confirm:

* I opened my existing WP Internal app, which I'm pretty sure I haven't uploaded any media from.
* I verified that both iOS system settings said my documents & data was 8.5MB and the app media cache said it was 8.5MB.
* I then uploaded a video (successfully).
* I then checked the numbers again: media cache, 8.5MB. Documents and data: 43.3MB!

Big issue!

Can you look into whether there's any way to get Alamofire to clear its cache more often? (i.e. after it's uploaded something successfully?!). Otherwise maybe we should clear it on app launch.

I like the idea of clearing it when clearing the device media cache, although I'm not sure if we should include the number there or not as I guess it's not 100% media in there?

**Comment by elibud** [comment]
What about configuring the cache to a reasonable size and clearing it when we clear the device media cache?

**Comment by frosty** [comment]
Looks like we might be able to set the cache through an NSURLSessionConfiguration: https://github.com/Alamofire/Alamofire/blob/master/Documentation/AdvancedUsage.md#session-manager

**Comment by SergioEstevao** [comment]
So I gave a deeper look at the way AlamoFire uploads files, specifically large files like media and I found the following:

 - When encoding files larger than 10Mb the system writes the file to a temporary directory 
 - It then encodes the request to that temporary directory using a file
 - It also add a file deletion task to be execute after the download is finished or there is an encoding error

So looking at the AlamoFire [code](https://github.com/Alamofire/Alamofire/blob/4.9.0/Source/SessionManager.swift#L711), the files used for the upload should be removed after an upload is done sucessfully or not, the only scenario were the file should stay there is if the upload is terminated by force close the app or stopping Xcode.

It's also strange for me that files stored in a folder obtained using NSTemporaryDirectory() count for the application size, I always thought those could be deleted by the OS when space is needed.

https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/FileSystemProgrammingGuide/FileSystemOverview/FileSystemOverview.html#//apple_ref/doc/uid/TP40010672-CH2-SW2

**Comment by frosty** [comment]
@SergioEstevao thanks for investigating!

> So looking at the AlamoFire code, the files used for the upload should be removed after an upload is done sucessfully or not, the only scenario were the file should stay there is if the upload is terminated by force close the app or stopping Xcode.

Interesting, as I was definitely seeing them remain there after a successful upload. I wonder if we should be regularly clearing it ourselves?

**Comment by SergioEstevao** [comment]
@frosty I will do a debug session to check if the code is executed or some error happens when trying to delete those.

BTW the URLCache you refer to in the SessionConfiguration is only used for the the responses we get from requests. By default is set to the [sharedURLCache](https://developer.apple.com/documentation/foundation/nsurlcache/1413377-sharedurlcache?language=objc) object that should have sensible value.

**Comment by SergioEstevao** [comment]
@frosty So I ran a debug session and I see the code for file removal to happen and the file to be removed. You checked the the file system and saw the file still present in the tmp folder? Or you just saw space being reported on the Settings section?

**Comment by SergioEstevao** [comment]
@frosty Remember that even after the video is uploaded we keep a local copy of the video file to use it if we are offline.

Should only do this for images? and don't keep any local cache for video Media?

**Comment by SergioEstevao** [comment]
So I run some more tests were I uploaded large files to a site, and the only scenario where files stay behind in the `org.alamofire.manager/multipart.form.data` folder is when the app killed in the middle of ongoing uploads.

This is looks like a bit of an edge case, but I can add code to clean that folder on app restart just to make sure those files are removed too.

> Actually, one more suggestion maybe — maybe we should include the size of tmp directory in the cache size displayed in the Settings and/or we should also clear it when the user taps on "Clear Device Media Cache"

One thing that I noticed is that our app media cache size only shows the space taken by thumbnails, and not by the local copies of media files we keep. I will put a PR up to add that size to the calculation for the number to be more realistic.
